src/test/scala/mesosphere/marathon/integration/UpgradeIntegrationTest.scala

```diff
@@ -0,0 +1,236 @@
+      marathon156.start().futureValue
+      (marathon156.client.info.entityJson \ "version").as[String] should be("1.5.6")
+
+      And("new apps in Marathon 1.5.6 are added")
+      val app_156 = appProxy(testBasePath / "app-156", "v1", instances = 1, healthCheck = None)
+      marathon156.client.createAppV2(app_156) should be(Created)
+
+      val app_156_fail = appProxy(testBasePath / "app-156-fail", "v1", instances = 1, healthCheck = None)
+      marathon156.client.createAppV2(app_156_fail) should be(Created)
+
+      Then("All apps from 1.5.6 are running")
+      eventually { marathon156 should have (runningTasksFor(app_156.id.toPath, 1)) }
+      eventually { marathon156 should have (runningTasksFor(app_156_fail.id.toPath, 1)) }
+
+      val originalApp156Tasks = marathon156.client.tasks(app_156.id.toPath).value
+      val originalApp156FailedTasks = marathon156.client.tasks(app_156_fail.id.toPath).value
+
+      And("All apps from 1.4.9 are still running")
+      marathon156.client.tasks(app_149.id.toPath).value should be(originalApp149Tasks)
+
+      When("Marathon 1.5.6 is shut down")
+      marathon156.stop().futureValue
+
+      And(s"App ${app_156_fail.id} fails")
+      killTask("app-156-fail")
+
+      // Pass upgrade to 1.6.322
+      And("Marathon is upgraded to 1.6.322")
+      marathon16322.start().futureValue
+      (marathon16322.client.info.entityJson \ "version").as[String] should be("1.6.322")
+
+      Then("All apps from 1.4.9 and 1.5.6 are still running")
+      marathon16322.client.tasks(app_149.id.toPath).value should be(originalApp149Tasks)
+      marathon16322.client.tasks(app_156.id.toPath).value should be(originalApp156Tasks)
+
+      // Pass upgrade to current
+      When("Marathon is upgraded to the current version")
+      marathon16322.stop().futureValue
+      marathonCurrent.start().futureValue
+      (marathonCurrent.client.info.entityJson \ "version").as[String] should be("1.6.0-SNAPSHOT")
+
+      Then("All apps from 1.4.9 and 1.5.6 are still running")
+      marathonCurrent.client.tasks(app_149.id.toPath).value should be(originalApp149Tasks)
+      marathonCurrent.client.tasks(app_156.id.toPath).value should be(originalApp156Tasks)
+
+      And("All apps from 1.4.9 and 1.5.6 are recovered and running again")
+      eventually { marathonCurrent should have(runningTasksFor(app_149_fail.id.toPath, 1)) }
+      marathonCurrent.client.tasks(app_149_fail.id.toPath).value should not be (originalApp149FailedTasks)
+
+      eventually { marathonCurrent should have(runningTasksFor(app_156_fail.id.toPath, 1)) }
+      marathonCurrent.client.tasks(app_156_fail.id.toPath).value should not be (originalApp156FailedTasks)
+
+      marathonCurrent.close()
+    }
+  }
+
+  def killTask(appName: String): Unit = {
+    val pidPattern = """([^\s]+)\s+([^\s]+)\s+.*""".r
+    val pids = Process("ps aux").!!.split("\n").filter { process =>
+      process.contains("src/test/python/app_mock.py") && process.contains(appName)
```

Comment:
What do you think about giving every app of this test an *additional* UUID in name and just checking `process.contains(uniqueAppName)` here?